crates/iceberg/src/transaction/rewrite_files.rs

```diff
@@ -0,0 +1,375 @@
+    ) -> Result<Self> {
+        for data_file in data_files {
+            match data_file.content {
+                DataContentType::Data => self.deleted_data_files.push(data_file),
+                DataContentType::PositionDeletes | DataContentType::EqualityDeletes => {
+                    self.deleted_delete_files.push(data_file)
+                }
+            }
+        }
+
+        Ok(self)
+    }
+
+    /// Set commit UUID for the snapshot.
+    pub fn set_commit_uuid(mut self, commit_uuid: Uuid) -> Self {
+        self.commit_uuid = Some(commit_uuid);
+        self
+    }
+
+    /// Set key metadata for manifest files.
+    pub fn set_key_metadata(mut self, key_metadata: Vec<u8>) -> Self {
+        self.key_metadata = Some(key_metadata);
+        self
+    }
+
+    /// Set snapshot summary properties.
+    pub fn set_snapshot_properties(mut self, snapshot_properties: HashMap<String, String>) -> Self {
+        self.snapshot_properties = snapshot_properties;
+        self
+    }
+
+    /// Set the data sequence number for this rewrite operation.
+    /// The number will be used for all new data files that are added in this rewrite.
+    pub fn set_data_sequence_number(mut self, sequence_number: i64) -> Self {
+        self.data_sequence_number = Some(sequence_number);
+        self
+    }
+
+    /// Set the snapshot ID used in any reads for this operation.
+    pub fn set_starting_snapshot_id(mut self, snapshot_id: i64) -> Self {
+        self.starting_snapshot_id = Some(snapshot_id);
+        self
+    }
+}
+
+impl Default for RewriteFilesAction {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+#[async_trait]
+impl TransactionAction for RewriteFilesAction {
+    async fn commit(self: Arc<Self>, table: &Table) -> Result<ActionCommit> {
+        let snapshot_producer = SnapshotProducer::new(
+            table,
+            self.commit_uuid.unwrap_or_else(Uuid::now_v7),
+            self.key_metadata.clone(),
+            self.snapshot_properties.clone(),
+            self.added_data_files.clone(),
+            self.added_delete_files.clone(),
+            self.deleted_data_files.clone(),
+            self.deleted_delete_files.clone(),
+        );
+
+        let rewrite_operation = RewriteFilesOperation {
+            added_data_files: self.added_data_files.clone(),
+            added_delete_files: self.added_delete_files.clone(),
+            deleted_data_files: self.deleted_data_files.clone(),
+            deleted_delete_files: self.deleted_delete_files.clone(),
+            starting_snapshot_id: self.starting_snapshot_id,
+            data_sequence_number: self.data_sequence_number,
+        };
+
+        // todo should be able to configure to use the merge manifest process
+        snapshot_producer
+            .commit(rewrite_operation, DefaultManifestProcess)
+            .await
+    }
+}
+
+fn copy_with_deleted_status(entry: &ManifestEntryRef) -> Result<ManifestEntry> {
```

Comment:
Why can't we just pass in a borrowed `ManifestEntry`? eg:

`copy_with_deleted_status(entry: &ManifestEntry)` ? `ManifestEntryRef` is an `Arc`, so callers that have an `Arc<ManifestEntry>` can pass in `entry.as_ref()`.

Reply:
Good point!